File: packages/claude-plugins/worktree/scripts/lib/ports.ts
import { createConnection } from 'node:net'

export function checkPort(port: number, host: string = '127.0.0.1'): Promise<boolean> {
    return new Promise((resolve) => {
        const socket = createConnection({ port, host })

        socket.once('connect', () => {
            socket.destroy()
            resolve(true) // Port is in use
        })

        socket.once('error', () => {
            socket.destroy()
            resolve(false) // Port is free
        })

        socket.setTimeout(1000, () => {
            socket.destroy()
            resolve(false) // Timeout = port likely free
        })
    })
}

export async function checkPorts(ports: number[]): Promise<Map<number, boolean>> {
    const results = new Map<number, boolean>()
    await Promise.all(
        ports.map(async (port) => {
            const inUse = await checkPort(port)
            results.set(port, inUse)
        })
    )
    return results
}

export function extractPortsFromSlot(slotConfig: Record<string, string | number>): number[] {
    const ports: number[] = []
    for (const [key, value] of Object.entries(slotConfig)) {
        if (key.toLowerCase().includes('port') && typeof value === 'number') {
            ports.push(value)
        }
    }
    return ports
}

File: packages/towles-tool/worktree/skills/lib/ports.py
"""Port availability checking."""
from __future__ import annotations

import socket
from typing import Any


def check_port(port: int, host: str = "127.0.0.1", timeout: float = 1.0) -> bool:
    """Check if port is in use. Returns True if in use."""
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.settimeout(timeout)
    try:
        sock.connect((host, port))
        sock.close()
        return True  # Connection succeeded = port in use
    except (socket.timeout, ConnectionRefusedError, OSError):
        return False  # Port is free


def check_ports(ports: list[int]) -> dict[int, bool]:
    """Check multiple ports. Returns dict of port -> in_use."""
    return {port: check_port(port) for port in ports}


def extract_ports_from_slot(slot_config: Any) -> list[int]:
    """Extract port values from slot config."""
    ports = []
    for key, value in slot_config.items():
        if "port" in key.lower() and isinstance(value, int):
            ports.append(value)
    return ports
